### DISCUSSION TITLE
Multi-threading : will operations on the same file handle always be called by the same thread as "open" ?

### DISCUSSION THREAD

**Discussion by UnasZole** [libfuse/libfuse#843] [Q&A]
Hi,

I have a bit of a specific question, which might sound silly to some, but I can't find a clear answer anywhere.

The FUSE documentation states that "For every open() call there will be exactly one release() call with the same flags and file handle", cf. https://github.com/libfuse/libfuse/blob/master/include/fuse.h#L529
And I know that one or several "flush" calls also occur in between.

My question is : when FUSE runs in multi-threaded mode, is there any guarantee that the flush and release calls on the same file handle will be performed by the same thread that did the "open" ?

The underlying question is : is it safe to store a mutex lock (in my case a std::shared_lock) within the file handle when "open" is invoked, to safely control which actions can or can't be done in parallel on a given file, and have that lock destroyed within the "release" call ?

**Comment by trapexit** [comment]
>  is there any guarantee that the flush and release calls on the same file handle will be performed by the same thread that did the "open" ?

No.

>  is it safe to store a mutex lock (in my case a std::shared_lock) within the file handle when "open" is invoked, to safely control which actions can or can't be done in parallel on a given file, and have that lock destroyed within the "release" call ?

Yes. Why wouldn't it be? The file handle is just a 64bit integer and typically people will be using that to store an address to some object you've allocated. How you choose to interact with that object is on you.

**Comment by UnasZole** [comment]
Thanks for the quick answer on the first question.

However, your second answer makes no sense to me, given the first.
A mutex lock is inherently linked to the specific thread that built it, as it materialises the ownership of the mutex by this thread.
The C++ documentation explicitly says it's undefined behaviour if a thread terminates while still owning a mutex (cf. https://en.cppreference.com/w/cpp/thread/mutex), and if a mutex is unlocked by another thread than the current owner (cf. https://en.cppreference.com/w/cpp/thread/mutex/unlock).

So if I have no guarantee that "release" will be called by the same thread as "open", it's absolutely not safe at all to store a mutex lock in the file handle.

**Reply by trapexit** [reply]
> A mutex lock is inherently linked to the specific thread that built it, as it materialises the ownership of the mutex by this thread.

That isn't what it says. If that were true then a mutex wouldn't be a mutex. It must be accessible by multiple threads. That's the whole point.

> The C++ documentation explicitly says it's undefined behaviour if a thread terminates while still owning a mutex .. and if a mutex is unlocked by another thread than the current owner 

And? That's entirely separate concern.

> So if I have no guarantee that "release" will be called by the same thread as "open", it's absolutely not safe at all to store a mutex lock in the file handle.

That's absolutely not how a mutex would be used. A mutex is used to safeguard critical sections. Not a possibly many hours long behavior you're describing. First, you are simply incorrect about the same thread needing to create and destroy. It needs to be the same thread that locks and then unlocks. Not thread A locking and thread B unlocking. Second... with that fact corrected you can create a mutex on open/create and have all functions which receive a file handle leverage said mutex to safeguard any critical sections. A release will only be issued when all other activities are finished so there is no race.

**Comment by bsbernd** [comment]
Storing a lock inside an oject and unlocking  from another thread are two different things... Doesn't make too much sense to me to lock the object and then returning to fuse-kernel/application.

**Comment by UnasZole** [comment]
> First, you are simply incorrect about the same thread needing to create and destroy

Sorry, but, no. It's just that you are misunderstanding what I'm saying - probably because you're not familiar with the C++ 11 API for concurrency ?
I'm not talking about storing a *mutex*, I've been asking about storing a *mutex lock*, the likes of std::unique_lock and std::shared_lock.
Those are objects applying the RAII principle to locking : when the lock is created, it locks the referenced mutex; once it's destroyed, it unlocks the mutex. And since locking and unlocking a mutex MUST be done from the same thread, as you said, then creating and destroying a lock must as well.

> Not a possibly many hours long behavior you're describing.

> Doesn't make too much sense to me to lock the object and then returning to fuse-kernel/application.

Well, you're both probably right. I'm not familiar enough with filesystem operations yet, and have no clear idea how long file descriptors typically stay open. Locking a mutex for such a long time is probably a poor idea indeed.


Let me explain what I'm trying to achieve : 
- My filesystem is in general a passthrough to a Iocal directory (heavily based on the passthrough_fh example), where all sorts of parallel operations are fine. (it has other specific functionalities, it's not "just" a passthrough, but let's keep it simple).
- But I have a background thread that will occasionally read files from this folder to copy them elsewhere. While this background thread is copying, I want a stable view of the file (ie I want to prevent other threads or applications to write to this file until the copy is complete).

My initial idea, probably wrong, was to assign a mutex per file, then a shared_lock to every file descriptor on that file in write mode, while the background thread would take a unique_lock on the same mutex. That way, it would wait for all "write" file descriptors to close before starting, and wouldn't give out any new write file descriptor until the copy is finished.

If the file descriptors can last several hours, that's obviously not going to scale well, though.
Is there a better way to achieve this behaviour ?

The question then being : when we open a write file descriptor and start writing to it, when do the changes become visible to all other open file descriptors on the same file ? Is it potentially immediate, or do the changes become committed only when "flush" is called ?
Because if so, I could probably just put my shared lock around the flush operation.

**Reply by bsbernd** [reply]
I would set a flag (like atomic variable) in the inode and let the background process continue on other files/inodes while that flag is set... No locking needed.

flush and visibility are not really related to each other. Assuming you would have a network file system, flush _might_ be needed to make changes visible (unless you write out the network server when a certain amount of data is reached, or you have a time writer, etc). Or if you would have a write cache and send writes to your underlying file system, etc. In your case it sounds like it is immediately visible,

**Reply by trapexit** [reply]
I seen no reason why you'd be storing a unique or shared lock beyond the scope of immediate use. That's not how they are supposed to be used. RAII requires stack allocation. The mutex would need to be allocated on the heap and used across all threads. And that's totally fine. The locks would only be used within the scope of the mutex's use.

Regardless, you don't have to understand the depths of filesystems to know that a client app can hold a file open for the entire lifetime of said app. Meaning it could be literally years.

POSIX has certain expectations on when data becomes available but every FS is a bit different and it is ultimately up to you to decide if your FS is even striving for compliance. Many / most FUSE filesystems are not fully.

I'm not sure I quite understand the purpose of all this. The best policy is to try to make any operation atomic so locking and waiting isn't necessary. If you must use locks then I you could probably do that strategy but you need to do that at the node level. A RW lock where the "writer" is the reader. But unless you are going to spawn a thread for every single file... blocking or spinning on a collection of mutexs (which is not a great idea)... then I'd suggest simply using atomic counters and polling. However, filesystems are inherently racy. Maybe step back and consider if what you're doing is proper in the first place or if you could leverage some other strategy to make some of this process more atomic.

**Reply by UnasZole** [reply]
Thanks both for the answers and detailed explanations.

> I'm not sure I quite understand the purpose of all this. The best policy is to try to make any operation atomic so locking and waiting isn't necessary. 

Well that's essentially what I'm trying to do, except that the operation I want to make "atomic" is pretty big.
Typically, my background thread has to do several things, including (and not limited to) computing a hash of the file and then copying the file to a remote location. I want this whole operation to become "atomic", ie to ensure the file is untouched during this whole process.
With the example I give here, you can easily understand that if the file is modified between the instant I compute the checksum and the instant I start transferring it, I'll have an inconsistency.
And ideally, I'd like to avoid the cost of copying the file to a temporary directory first, and then only acting on the copy.

(Not to mention that I'd ideally like to avoid triggering this background thread if the file is still in the middle of being edited, which is why I was initially thinking of locking at "open" and unlocking at "release", instead of locking at "write".)

Hence why forcing the writes to wait while the background thread is working on the file seemed the best strategy.
If you have other ideas that don't involve locking at all, I'm interested of course.


In the meantime, @bsbernd suggestion made me think of a different approach, which would be the following : 
- Store an atomic counter for each file, counting the number of 'write' file descriptors on it. On "open", increment the counter ; on "release" decrement it.
Then, my background thread ignores files where this counter is non-zero.
This takes care of the first half of the problem : preventing the background thread to act on files which have an open "write" file descriptor (ie files on which data may be written at any time).
- Also store a shared_mutex for each file. The background thread takes an exclusive lock on it, while the "open" method takes a shared lock before instantiating and returning the file descriptor. (the lock is destroyed when "open" returns)
This takes care of the second half : ensuring new "write" file descriptors are not created while the background thread is processing.

Does that sound like a safe way to ensure the file won't be edited while the background thread processes it ?

(Of course, in that case, files which stay constantly open for hours or days will not be picked up by the background thread, but for the time being it's a limitation I can live with.)

**Reply by trapexit** [reply]
The problem I see is just that "middle of being edited" is a human, high level idea. Software can (and do) all kinds of patterns of access depending on the whim of the author. If you know your users are going to work in a certain way then you can account for that but if you expect random software to work it may not. Take torrent apps.... they regularly create files with lots of holes in them and to keep file descriptor count down will "open, write, close" or "open, read, close" pretty regularly. Sometimes they have a LRU open file cache but still... that kind of usage would thrash your background thread and/or cause major blockage on the file.

**Reply by UnasZole** [reply]
Thanks @trapexit ! Even though I'm really not designing that filesystem for use with torrent apps, this example made it pretty clear that "file descriptor" is not the correct level to perform any kind of locking.
I was working on the misconception that changes were only committed in the files when "flush" was called on the FD, but since that's not the case (as @bsbernd pointed out), then the correct level for locking in order to ensure a consistent view on the file is the "write" operation.

Then, I need another separate mechanism to avoid my background thread acting "in the middle of editing", which would by essence not be 100% reliable but would reduce the number of useless actions. Possibly just have the thread ignore files which have a filemtime within the last x seconds or minutes (if the file has not been touched for a while, it's more likely to be in a "stable state" worth considering).

By the way, what is the usual behaviour of filemtime ? Is it updated on every "write" call on a file ?

**Reply by trapexit** [reply]
I wouldn't use mtime either. You still have a race condition on usage. As mentioning before you want to boil things down to atomic pieces or as close as you can. Writes come in with a payload. You can take that write request and queue it up to be worked on. If you want to put a delay to batch things up then you can but you could still get into a situation where you need to rescan the list due to a change mid file.

POSIX compatible remote filesystems are hard. The posix filesystem api was not designed for high latency, high concurrent behavior.

**Comment by Nikratio** [comment]
It seems to that you do not need any sort of per-file lock. What you need is a data structure that keeps track of while file is currently being checksummed, and one mutex to protect access to that.

So, use a `std::unordered_set` to keep track of what's being hashed, and protect access to that by a single `std::condition_variable`. On write, obtain the lock and check if you are allowed to write. If so, go ahead and then release the lock. If you're not allowed to write, use the condition variable to get woken up when the set changes and recheck.

(If you're worried about serializing writes this way, you can use another set to keep track of what's currently being written to get back parallelism).

**Reply by UnasZole** [reply]
Thanks a lot for the suggestion !
To handle locking per-file I was focused on having one mutex per file, but I might be able to get away with a single condition variable associated with a container describing what resources are currently locked. I didn't think of that :-)
I'll check if I can integrate that idea, that could significantly simplify my setup.
